### PR TITLE
New engine proposal - Multi stream & simpler dispatcher/reducer

### PR DESCRIPTION
**Pull request by arnaudgiuliani**
Hi all 👋 

TL;RD

- all changes are available in `0.13.0-alpha-1`
- keep mainly all current API
- allow to use multiple state/events streams
- reworked internals to avoid use coroutines Flow as internal plumbing

⚠️ long thread ⚠️

Here is a proposal about reworking the main core components. the idea: allow work on a default/main data publisher as many data publishers you want. Keep the easy way to write actions, but with more streams.

The main concept remains the same: the action allows to ensure a synchrone update of your ViewModel data streams, thanks to the action dispatcher. An action will be executed and then will update the following data publishers.

It's also a good time to simplify core components:
- ActionDispatcher, to just create the Action object to pass for reduce
- ActionReducer, to enqueue an action & run/reduce it to update new states
- DataPublisher to help have default publisher (classical mono stream ViewModel) or write more complex ones 

ActionReducer avoids the use of Flow internally as event passing pipeline. 

What matters in all of that, is the execution of actions one by one. 

This new design keeps the current API as is 👍 (action { setState( ...) } , actionOn<>)

One thing I didn't keep: The BadOrWrong state guard is now emitting an exception rather than a custom event. It's more convenient to let the developers handle it as they want.

--

Here below the main components rewritten:

DataFlow:
```kotlin
interface DataFlow {
    val tag: String
    val actionDispatcher: ActionDispatcher
    fun defaultPublisher(): DataPublisher
    fun action(onAction: ActionFunction): Action = actionDispatcher.dispatchAction(onAction)
    fun action(onAction: ActionFunction, onError: ActionErrorFunction): Action = actionDispatcher.dispatchAction(onAction, onError)
    suspend fun onError(error: Exception, currentState: UIState) {
        UniFlowLogger.logError("Uncaught error: $error - ${error.stackTrace}", error)
        throw error
    }
}

inline fun <reified T : UIState> DataFlow.actionOn(noinline onAction: ActionFunction_T<T>): Action = actionDispatcher.actionOn(T::class, onAction as ActionFunction)
inline fun <reified T : UIState> DataFlow.actionOn(noinline onAction: ActionFunction_T<T>, noinline onError: ActionErrorFunction_T<T>): Action = actionDispatcher.actionOn(T::class, onAction as ActionFunction, onError as ActionErrorFunction)
```

Action:
```kotlin
typealias ActionFunction = suspend (UIState) -> (Unit)
typealias ActionFunction_T<T> = suspend (T) -> (Unit)
typealias ActionErrorFunction = suspend (Exception, UIState) -> (Unit)
typealias ActionErrorFunction_T<T> = suspend (Exception, T) -> (Unit)

class Action(
        val onSuccess: ActionFunction,
        val onError: ActionErrorFunction,
        val targetState : KClass<out UIState>? = null
)
```

ActionDispatcher:
```kotlin
class ActionDispatcher(
        private val coroutineScope: CoroutineScope,
        private val reducer: ActionReducer,
        private val runError: suspend (Exception, UIState) -> Unit,
        val tag: String
) {
    fun dispatchAction(onAction: ActionFunction): Action = dispatchAction(onAction) { error, state -> runError(error, state) }

    fun dispatchAction(onAction: ActionFunction, onError: ActionErrorFunction): Action = Action(onAction, onError).also {
        dispatchAction(it)
    }

    fun dispatchAction(action: Action) {
        coroutineScope.launchOnIO {
            UniFlowLogger.debug("$tag - enqueue: $action")
            reducer.enqueueAction(action)
        }
    }

    fun actionOn(kClass: KClass<out UIState>, onAction: ActionFunction): Action = actionOn(kClass,onAction){ error, state -> runError(error, state) }
    fun actionOn(kClass: KClass<out UIState>, onAction: ActionFunction, onError: ActionErrorFunction): Action = Action(onAction, onError, kClass).also { dispatchAction(it) }

    fun close() {
        reducer.close()
    }
}
```

ActionReducer:
```kotlin
class ActionReducer(
        private val defaultPublisher: () -> DataPublisher,
        private val coroutineScope: CoroutineScope,
        private val defaultDispatcher: CoroutineDispatcher,
        defaultCapacity: Int = Channel.BUFFERED,
        private val tag: String
) {

    @OptIn(ObsoleteCoroutinesApi::class)
    private val actor = coroutineScope.actor<Action>(UniFlowDispatcher.dispatcher.default(), capacity = defaultCapacity) {
        for (action in channel) {
            if (coroutineScope.isActive) {
                withContext(defaultDispatcher) {
                    reduceAction(action)
                }
            } else {
                UniFlowLogger.debug("$tag - $action cancelled")
            }
        }
    }

    suspend fun enqueueAction(action: Action) {
        actor.send(action)
    }

    private suspend fun reduceAction(action: Action) {
        UniFlowLogger.debug("$tag - reduce: $action")
        val currentState: UIState = defaultPublisher().getState()
        try {
            action.targetState?.let { targetState ->
                if (targetState != currentState::class) {
                    action.onError(BadOrWrongStateException(currentState, targetState), currentState)
                    return
                }
            }
            action.onSuccess(currentState)
            UniFlowLogger.debug("$tag - completed: $action")
        } catch (e: Exception) {
            UniFlowLogger.debug("$tag - error: $action")
            action.onError(e, currentState)
        }
    }

    fun close() {
        actor.close()
    }
}
```

DataPublisher:
```kotlin
interface StatePublisher {
    suspend fun publishState(state: UIState, pushStateUpdate: Boolean = true)
    suspend fun getState(): UIState
    suspend fun setState(state: UIState) = publishState(state)
    suspend fun setState(state: () -> UIState) = publishState(state())
}

interface EventPublisher {
    suspend fun publishEvent(event: UIEvent)
    suspend fun sendEvent(event: UIEvent) = publishEvent(event)
    suspend fun sendEvent(event: () -> UIEvent) = publishEvent(event())
}

interface DataPublisher : StatePublisher, EventPublisher {
    suspend fun notifyStateUpdate(state: UIState, event: UIEvent) {
        publishState(state, pushStateUpdate = false)
        publishEvent(event)
    }
}

suspend inline fun <reified T : UIState> StatePublisher.getStateOrNull(): T?{
    val state = getState()
    return if (state is T){
        state as? T
    } else null
}
```

-- The Android impl --

```kotlin
abstract class AndroidDataFlow(
        defaultState: UIState = UIState.Empty,
        defaultCapacity: Int = Channel.BUFFERED,
        defaultDispatcher: CoroutineDispatcher = UniFlowDispatcher.dispatcher.io()
) : ViewModel(), DataFlow, DataPublisher {
    override val tag = this.toString()

    private val coroutineScope: CoroutineScope = viewModelScope

    val defaultDataPublisher = liveDataPublisher(defaultState, "main")
    override suspend fun publishState(state: UIState, pushStateUpdate: Boolean) = defaultPublisher().publishState(state, pushStateUpdate)
    override suspend fun publishEvent(event: UIEvent) = defaultPublisher().publishEvent(event)
    override suspend fun getState(): UIState = defaultPublisher().getState()
    override fun defaultPublisher(): DataPublisher = defaultDataPublisher

    private val actionReducer = ActionReducer(::defaultPublisher, coroutineScope, defaultDispatcher, defaultCapacity, tag)
    override val actionDispatcher: ActionDispatcher = ActionDispatcher(coroutineScope, actionReducer, ::onError, tag)

    @CallSuper
    override fun onCleared() {
        actionReducer.close()
        super.onCleared()
    }
}
```

```kotlin
class LiveDataPublisher(defaultState: UIState, val tag: String) : DataPublisher {

    private val _states = MutableLiveData<UIState>()
    val states: LiveData<UIState> = _states
    private val _events = MutableLiveData<Event<UIEvent>>()
    val events: LiveData<Event<UIEvent>> = _events

    init {
        _states.value = defaultState
    }

    override suspend fun getState(): UIState = _states.value ?: error("No state in LiveData")
    override suspend fun publishState(state: UIState, pushStateUpdate: Boolean) {
        onMain(immediate = true) {
            UniFlowLogger.debug("$tag <-- $state")
            _states.value = state
        }
    }

    override suspend fun publishEvent(event: UIEvent) {
        onMain(immediate = true) {
            UniFlowLogger.debug("$tag <-- $event")
            _events.value = Event(content = event)
        }
    }
}

fun liveDataPublisher(defaultState: UIState, tag: String) = LiveDataPublisher(defaultState,tag)
```

### CODE REVIEW

**Comment by arnaudgiuliani** [comment]
Some code sample here with this new "engine":

```kotlin
class MultiFlow(private val repository: TodoRepository) : AbstractSampleFlow(UIState.Empty) {

    val pub1 = simpleListPublisher(UIState.Empty, "#1")
    val pub2 = simpleListPublisher(UIState.Empty, "#2")

    fun addMulti(todo: String) = action {
        repository.add(todo)
        val todos = repository.getAllTodo()

        println("new list -> ${todos.size}")
        pub1.setState { todos.mapToTodoListState() }
        pub2.setState { CountState(todos.size) }
    }
```
and then we can test it:

```kotlin
@Test
    fun `simple multi action`() {
        dataFlow.addMulti("toto")
        val allTodo = repository.getAllTodo()
        dataFlow.assertReceived(
                listOf(
                        UIState.Empty,
                        allTodo.mapToTodoListState()
                ),
                listOf(
                        UIState.Empty,
                        CountState(allTodo.size)
                )
        )
    }
```

File: uniflow-androidx/src/main/java/io/uniflow/androidx/flow/AndroidDataFlow.kt [hunk 1]
```diff
@@ -44,25 +47,23 @@ abstract class AndroidDataFlow(
...
-    private val actionDispatcher: ActionDispatcher
-        get() = ActionDispatcher(viewModelScope, reducer, dataStore, this,tag)
 
-    final override fun getCurrentState() = actionDispatcher.getCurrentState()
-    final override fun <T : UIState> getCurrentStateOrNull(stateClass: KClass<T>): T? = actionDispatcher.getCurrentStateOrNull()
-    final override fun action(onAction: ActionFunction<UIState>): ActionFlow = actionDispatcher.action(onAction)
-    final override fun action(onAction: ActionFunction<UIState>, onError: ActionErrorFunction): ActionFlow = actionDispatcher.action(onAction, onError)
-    final override fun <T : UIState> actionOn(stateClass: KClass<T>, onAction: ActionFunction<T>): ActionFlow = actionDispatcher.actionOn(stateClass, onAction)
-    final override fun <T : UIState> actionOn(stateClass: KClass<T>, onAction: ActionFunction<T>, onError: ActionErrorFunction): ActionFlow = actionDispatcher.actionOn(stateClass, onAction, onError)
+    val defaultDataPublisher = liveDataPublisher(defaultState, "main")
```
Comment (MarcinChrapowicz) [inline uniflow-androidx/src/main/java/io/uniflow/androidx/flow/AndroidDataFlow.kt:55]: Can we have a const for `"main"`  also we can remove `default` from  `defaultDataPublisher`, using just `dataPublisher` should be fine :)

File: uniflow-androidx/src/main/java/io/uniflow/androidx/flow/AndroidDataFlow.kt [hunk 2]
```diff
@@ -44,25 +47,23 @@ abstract class AndroidDataFlow(
...
+    override suspend fun publishState(state: UIState, pushStateUpdate: Boolean) = defaultPublisher().publishState(state, pushStateUpdate)
+    override suspend fun publishEvent(event: UIEvent) = defaultPublisher().publishEvent(event)
+    override suspend fun getState(): UIState = defaultPublisher().getState()
+    override fun defaultPublisher(): DataPublisher = defaultDataPublisher
```
Comment (MarcinChrapowicz) [inline uniflow-androidx/src/main/java/io/uniflow/androidx/flow/AndroidDataFlow.kt:59]: Same here why we are adding `default` at the beginning ?, do we have more options of `DataPublisher` ?? 
maybe we can just use name `publisher`

File: uniflow-core/src/main/kotlin/io/uniflow/core/flow/DataFlow.kt [hunk 3]
```diff
@@ -13,30 +13,29 @@
...
-    fun getCurrentState(): UIState
-    fun <T : UIState> getCurrentStateOrNull(stateClass: KClass<T>): T?
-    fun action(onAction: ActionFunction<UIState>): ActionFlow
-    fun action(onAction: ActionFunction<UIState>, onError: ActionErrorFunction): ActionFlow
-    fun <T : UIState> actionOn(stateClass: KClass<T>, onAction: ActionFunction<T>): ActionFlow
-    fun <T : UIState> actionOn(stateClass: KClass<T>, onAction: ActionFunction<T>, onError: ActionErrorFunction): ActionFlow
-    suspend fun onError(error: Exception, currentState: UIState, flow: ActionFlow) {
+    val tag: String
+    val actionDispatcher: ActionDispatcher
+    fun defaultPublisher(): DataPublisher
```
Comment (MarcinChrapowicz) [inline uniflow-core/src/main/kotlin/io/uniflow/core/flow/DataFlow.kt:31]: like above we can use `publisher`

File: uniflow-test/src/test/kotlin/io/uniflow/test/impl/AbstractSampleFlow.kt [hunk 4]
```diff
@@ -1,47 +1,41 @@
...
-    private val actionDispatcher: ActionDispatcher
-        get() = ActionDispatcher(coroutineScope, reducer, dataStore, this, tag)
 
-    final override fun getCurrentState() = actionDispatcher.getCurrentState()
-    final override fun <T : UIState> getCurrentStateOrNull(stateClass: KClass<T>): T? = actionDispatcher.getCurrentStateOrNull()
-    final override fun action(onAction: ActionFunction<UIState>): ActionFlow = actionDispatcher.action(onAction)
-    final override fun action(onAction: ActionFunction<UIState>, onError: ActionErrorFunction): ActionFlow = actionDispatcher.action(onAction, onError)
-    final override fun <T : UIState> actionOn(stateClass: KClass<T>, onAction: ActionFunction<T>): ActionFlow = actionDispatcher.actionOn(stateClass, onAction)
-    final override fun <T : UIState> actionOn(stateClass: KClass<T>, onAction: ActionFunction<T>, onError: ActionErrorFunction): ActionFlow = actionDispatcher.actionOn(stateClass, onAction, onError)
+    internal val defaultDataPublisher = simpleListPublisher(defaultState, "main")
```
Comment (MarcinChrapowicz) [inline uniflow-test/src/test/kotlin/io/uniflow/test/impl/AbstractSampleFlow.kt:22]: You can apply the same changes here. From -> https://github.com/uniflow-kt/uniflow-kt/pull/56/files#diff-20ab9838c64e7a207e75b2d7617f4c5e336e6c81dd4233cf6d6126fde25eef15

**Comment by MarcinChrapowicz** [comment]
Nice work 👏 , just one thing that is missing, maybe we should add some information in the `README` about that?

**Comment by erikhuizinga** [comment]
I'm going to unsub from this PR because of time/prio conflicts. From what I've seen this is a nice development.

**Comment by arnaudgiuliani** [comment]
@MarcinChrapowicz I'm putting `default` publisher in `DataFlow` because it's the first default publisher. If you don't define any, there is a default one :)

File: uniflow-androidx/src/main/java/io/uniflow/androidx/flow/AndroidDataFlow.kt (hunk 1 above)
Comment (arnaudgiuliani) [reply uniflow-androidx/src/main/java/io/uniflow/androidx/flow/AndroidDataFlow.kt:55]: yes, "main" should be an internal constant. For `defaultDataPublisher` naming, I begin to reply in general comments https://github.com/uniflow-kt/uniflow-kt/pull/56#issuecomment-743046553

**Comment by MarcinChrapowicz** [comment]
I see, fell free to merge it :D

File: gradle/versions.gradle [hunk 5]
```diff
@@ -1,10 +1,10 @@
 ext {
     // Koin
-    uniflow_version = '0.12.0-beta-1'
+    uniflow_version = '0.13.0-alpha-1'
 
     // Kotlin
-    kotlin_version = '1.3.72'
-    coroutines_version = "1.3.5"
```
Comment (costular) [inline gradle/versions.gradle:7]: You could update kotlin to `1.4.21`

File: gradle/versions.gradle [hunk 6]
```diff
@@ -1,10 +1,10 @@
...
+    kotlin_version = '1.4.10'
```
Comment (costular) [inline gradle/versions.gradle:6]: You could update this library to `1.4.2`

**Review by costular — COMMENTED** [review]
In general it looks good to me. I'd just update these libraries. Nice work 👍🏼

File: gradle/versions.gradle (hunk 5 above)
Comment (arnaudgiuliani) [reply gradle/versions.gradle:7]: yep 👍

**Comment by arnaudgiuliani** [comment]
> In general it looks good to me. I'd just update these libraries. Nice work 👍🏼

great, thanks :)

**Pull request merged**